rp2040-hal/src/adc.rs

```diff
@@ -153,3 +154,210 @@ where
         Ok(self.device.result.read().result().bits().into())
     }
 }
+
+impl Adc {
+    pub fn free_running<'a>(&'a mut self) -> FreeRunning<'a> {
+        FreeRunning { adc: self }
+    }
+}
+
+pub struct FreeRunning<'a> {
+    adc: &'a mut Adc,
+}
+
+impl<'a> FreeRunning<'a> {
+    /// Manually set clock divider integral and fractional parts
+    pub fn clock_divider(self, int: u16, frac: u8) -> Self {
+        self.adc.device.div.modify(|_, w| unsafe { w.int().bits(int).frac().bits(frac) });
+        self
+    }
+
+    /// Select initial ADC input to sample from
+    ///
+    /// If round-robin mode is used, this will only affect the first sample.
+    pub fn initial_input<PIN: Channel<Adc, ID = u8>>(self, _pin: &mut PIN) -> Self {
+        self.adc.device.cs.modify(|_, w| unsafe { w.ainsel().bits(PIN::channel()) });
+        self
+    }
+
+    /// Set channels to use for round-robin mode
+    ///
+    /// Takes a tuple of channels, like `(&mut adc_pin, &mut temp_sense)`.
+    ///
+    /// The order in which the channels are specified has no effect.
+    /// Channels are always sampled in increasing order (Channel 0, Channel 1, ...).
+    pub fn round_robin<T: Into<RoundRobin>>(self, selected_channels: T) -> Self {
+        let RoundRobin(bits) = selected_channels.into();
+        self.adc.device.cs.modify(|_, w| unsafe { w.rrobin().bits(bits) });
+        self
+    }
+
+    /// Enable ADC fifo and start free-running conversion
+    ///
+    /// Use the returned `Fifo` instance to access the captured data.
+    ///
+    /// To stop capturing, call `fifo.stop()`.
+    pub fn start_fifo(self) -> Fifo<'a, false> {
+        self.adc.device.fcs.modify(|_, w| w.en().set_bit());
+        self.adc.device.cs.modify(|_, w| w.start_many().set_bit());
+        Fifo { adc: self.adc }
+    }
+
+    /// Enable ADC fifo with DMA and start free-running conversion
+    ///
+    /// In addition to being usable the same way as a `Fifo` returned from `start_fifo`,
+    /// the Fifo returned by this function can also be used as a source for DMA transfers.
+    pub fn start_fifo_with_dma(self, thresh: u8) -> Fifo<'a, true> {
```

Comment:
The datasheet says: "The threshold for DREQ assertion (FCS.THRESH) should be set to 1, so that the DMA transfers as soon as a single sample is present in the FIFO. Note this is also the threshold used for IRQ assertion, so non-DMA use cases might prefer a higher value for less frequent interrupts."

Therefore I'd say for the DMA use case, the threshold should not be configurable, but fixed to 1.
Also, it probably (but see my other comment) doesn't harm to always enable DREQ. So `start_fifo_with_dma` and `start_fifo` may not need to be separate functions? And the DmaEnabled flag could be removed from the Fifo type signature?

Reply:
hm, yeah, those are some good points.

If DREQ is always enabled, we can't enforce FCS.THRESH=1 for the DMA usecase (or else not have the threshold configurable).

We could move this issue to the user (at least until it's clear how to handle it nicely API-wise), by having builder methods `set_threshold(...)`, `enable_dma(true | false)`, `enable_interrupt(true | false)`, and have a single `start_fifo` function as you suggest.
That would make these features usable without messing with registers manually, but it would leave some caveats:
- either there needs to be some note, that when enabling DMA, the threshold should be set to 1
- or `enable_dma` would force the threshold to 1, possibly overriding the user choice (or not, if the user calls `set_threshold` after `enable_dma`)

Not the best options, imho.

So, also in light of your other comments: maybe the DMA part can be skipped for now, and handled in a future PR, when more is known about the usecases.
Then my goal would be to make the API nice for the interrupt usecase.

Which makes me wonder: do you think there is a usecase where neither the interrupt, nor DMA is used?
From what I can tell when capturing in free-running mode, CS.READY does not change, so even though the samples can be read from CS.RESULT there is no way to tell when a new sample was captured.

Reply:
I agree, that sounds reasonable.

Since `AdcFifo` currently holds a `&mut Adc` calling `adc.read_single()` directly wouldn't work though (unless the `AdcFifo` is dropped first, which makes it impossible to stop the conversion later on).
So I've added a `read_single` method to `AdcFifo` now, which just calls the underlying `adc.read_single` to enable this use case.